Crossfire: add passthrough telemetry over crossfire

yaapu opened the pull request:
this add passthrough telemetry using crossfire as a transport.
Main features:
- supports both legacy crossfire telemetry along passthrough at the same time
- supports both rf mode 1 (slow) and rf mode 2 (fast)
- uses a dedicated frame for status text messages
- when in rf mode 1 packs multiple passthrough packets in a single big frame to save bandwidth
- detects crossfire fw version to get supported features

More info [here](https://discuss.ardupilot.org/t/passthrough-telemetry-over-crsf-crossfire/62668)

peterbarker commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -116,6 +119,13 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
     };
 
+    // SubType IDs for CRSF_FRAMETYPE_CUSTOM_TELEM
+    enum CustomTelemSubTypeID : uint8_t {
+        CRSF_CUSTOM_TELEM_PASSTHROUGH = 0xF0,
+        CRSF_CUSTOM_TELEM_STATUS_TEXT = 0xF1,
+        CRSF_CUSTOM_TELEM_PASSTHROUGH_ARRAY = 0xF2,
+    };
```
preexisting, but these are really ugly... should be:
```suggestion
    enum class CustomTelemSubType : uint8_t {
        PASSTHROUGH = 0xF0,
        STATUS_TEXT = 0xF1,
        PASSTHROUGH_ARRAY = 0xF2,
    };
```

And users go from `CRSF_CUSTOM_TELEM_PASSTHROUGH` to `CustomTelemSubType::PASSTHROUGH`.

Not for this PR, 'though.

yaapu replied:
yeah, probably best to do it in a separate PR where we refactor all similar enums

andyp1per replied:
@peterbarker I really don't agree with the use of enum class where you are representing numbers that actually go on the wire - it's far to inconvenient and adds little value IMO other than lots of casting.

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    return AP_RCProtocol_CRSF::CRSF_FRAMETYPE_CUSTOM_TELEM_LEGACY;
+}
+
+uint8_t AP_CRSF_Telem::get_rf_mode() const
```
```suggestion
RF_Mode AP_CRSF_Telem::get_rf_mode() const
```

yaapu replied:
see rf_mode comment

yaapu replied:
done

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+{
+    AP_RCProtocol_CRSF* crsf = AP::crsf();
+    if (crsf == nullptr) {
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ;
```
Should there really be a `NONE` entry in that enumeration?

yaapu replied:
don't know really, the radio will switch to whatever it can based on link quality, NONE might mean not detected yet, but what do we do if we did not detect it?,My assumption was I go for 150Hz and eventaully detect it later

yaapu replied:
perhaps better an UNKNOWN, done

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+
+    if (_scheduler.avg_packet_rate > 55) {  // need to carefully select this value!!
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ;
+    } else {
```
Eliminate this else

yaapu replied:
fixed

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+            } else {
+                _request_pending.destination = AP_RCProtocol_CRSF::CRSF_ADDRESS_CRSF_RECEIVER;
+                _request_pending.frame_type = AP_RCProtocol_CRSF::CRSF_FRAMETYPE_PARAM_DEVICE_INFO;
+                gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: sending RX device ping");
```
```suggestion
                gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: requesting RX device ping");
```

peterbarker replied:
What limits the rate we send these requests at?

yaapu replied:
The AP_CRSF_Telem scheduler, it will have a slot for parameters every once in a while when requested to feed data to AP_RCProtocol_CRSF

yaapu replied:
aren't we technically sending a ping and requesting device info?... i refactored the status text but not 100% convinced :-)

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -245,7 +376,45 @@ void AP_CRSF_Telem::process_ping_frame(ParameterPingFrame* ping)
+        uint8_t Parameters count
+        uint8_t Parameter version number        
+    */
+    const uint8_t offset = MIN(strlen((char*)info->payload), 42);
```
What guarantees null-termination on payload here?  We do not want to be walking off the end of that buffer looking for a null char
.

yaapu replied:
nothing in reality, a proper implementation in the crossfire firmware I guess, I might check the frame size though

yaapu replied:
I'll switch to a loop to detect that

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -268,13 +437,16 @@ void AP_CRSF_Telem::update()
         calc_parameter();
         break;
+    case AP_RCProtocol_CRSF::CRSF_FRAMETYPE_PARAM_DEVICE_INFO:
+        calc_device_ping();
```
It seems rather odd that we calculate pin under the device-info case and vice-versa.

yaapu replied:
@andyp1per yes it seemed odd to me too but I respected the pre-existing approach

andyp1per replied:
It's request-response - you are pinged and you return a device info. If you are handed a device info, not sure why you then need a ping, the ping should have already happened.

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1011,80 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+    if (!_statustext.available) {
+        WITH_SEMAPHORE(_statustext.sem);
+        // check link speed
+        if (_telem_rf_mode != AP_RCProtocol_CRSF::CRSF_RF_MODE_150HZ) {
```
Perhaps a `bool high_speed_telemetry() const` call might be useful as we do this check in multiple places.

yaapu replied:
agree, done

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1011,80 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+            if (!got_message) {
+                return;
+            }
+        } else {
+            if (!_statustext.queue.pop(_statustext.next)) {
```
```suggestion
        } else if (!_statustext.queue.pop(_statustext.next)) {
```

yaapu replied:
fixed

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1011,80 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+    _telem_type = get_custom_telem_frame_id();
+    _telem.bcast.custom_telem.status_text.sub_type = AP_RCProtocol_CRSF::CustomTelemSubTypeID::CRSF_CUSTOM_TELEM_STATUS_TEXT;
+    _telem.bcast.custom_telem.status_text.severity = _statustext.next.severity;
+    const uint8_t len = hal.util->snprintf(_telem.bcast.custom_telem.status_text.text, 50, "%s", _statustext.next.text);
```
ARRAY_SIZE rather than 50

peterbarker replied:
`len` here is *NOT* the number of characters written into `_telem.bcast.custom_telem.status_text.text`.

It is the number of characters `snprintf` would have written if it was able to write everything.

You may wish to re-jig all of this using `strncpy` instead...

yaapu replied:
ok done

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1011,80 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+    uint8_t count = 0;
+    AP_Frsky_SPort::sport_packet_t buffer[9] {};
+    // we request a 9 packet array
+    if (!AP_Frsky_Telem::get_telem_data_array(buffer, count, 9)) {
```
```suggestion
    if (!AP_Frsky_Telem::get_telem_data_array(buffer, count, ARRAY_SIZE(buffer))) {
```

yaapu replied:
fixed

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -213,49 +246,76 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
      // setup ready for passthrough operation
     void setup_wfq_scheduler(void) override;
-
+    
```
Stray whitespace.

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -213,49 +246,76 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
+    struct {
+        uint8_t destination = AP_RCProtocol_CRSF::CRSF_ADDRESS_BROADCAST;
+        uint8_t frame_type;
+    } PACKED _request_pending;
```
Why is this packed?

Also, shouldn't this be `_pending_request` rather than `_request_pending`?

yaapu replied:
yep...better

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -213,49 +246,76 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
+        uint8_t major;
+        uint8_t retry_count;
+        bool use_rf_mode;
+        bool pending = true;
```
Should this really be pending as a default?

yaapu replied:
yes, since at boot I need to get the version to complete custom telemetry initialization.
I could save some by not initializing it here but would need to set it to true somewhere during initialization, no idea about the impact.

peterbarker commented on libraries/AP_RCTelemetry/AP_RCTelemetry.cpp:
```diff
@@ -84,7 +84,7 @@ void AP_RCTelemetry::update_avg_packet_rate()
  * WFQ scheduler
  */
-void AP_RCTelemetry::run_wfq_scheduler(void)
+uint8_t AP_RCTelemetry::run_wfq_scheduler(const bool use_shaper)
```
Please comment on what the return value of this function is.

yaapu replied:
ok done

peterbarker commented on libraries/AP_RCTelemetry/AP_RCTelemetry.cpp:
```diff
@@ -116,8 +116,9 @@ void AP_RCTelemetry::run_wfq_scheduler(void)
-        if (delay >= max_delay && ((now - _scheduler.packet_timer[i]) >= _scheduler.packet_min_period[i])) {
-            packet_ready = is_packet_ready(i, queue_empty);
+        if (delay >= max_delay && check_scheduler_entry_time_constraints(now, i, use_shaper)) {
+            // check if the entry is disabled and if not call is_packet_ready
```
Comment is redundant.

yaapu replied:
removed

peterbarker commented on libraries/AP_RCTelemetry/AP_RCTelemetry.h:
```diff
@@ -36,21 +36,58 @@ class AP_RCTelemetry {
+        }
+        _scheduler.packet_min_period[slot] = min_period_ms;
+    }
+    bool is_scheduler_entry_enabled(const uint8_t slot) {
```
```suggestion
    bool is_scheduler_entry_enabled(const uint8_t slot) const {
```

yaapu replied:
fixed

peterbarker commented on libraries/AP_RCTelemetry/AP_RCTelemetry.h:
```diff
@@ -81,16 +118,22 @@ class AP_RCTelemetry {
     virtual bool is_packet_ready(uint8_t idx, bool queue_empty) { return true; }
     virtual void process_packet(uint8_t idx) = 0;
     virtual void adjust_packet_weight(bool queue_empty) {};
+    bool check_scheduler_entry_time_constraints(const uint32_t now, uint8_t slot, const bool use_shaper) {
```
```suggestion
    bool check_scheduler_entry_time_constraints(const uint32_t now, uint8_t slot, const bool use_shaper) const {
```

yaapu replied:
fixed

peterbarker commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -395,6 +395,11 @@ class RC_Channels {
         return get_singleton() != nullptr && (_options & uint32_t(Option::FPORT_PAD));
     }
 
+    // should we use passthrough data forr crsf telemetry
```
```suggestion
    // returns true if we should pass through data for crsf telemetry
```

yaapu replied:
done

peterbarker commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -442,14 +447,15 @@ class RC_Channels {
-        ARMING_CHECK_THROTTLE = (1 << 5), // run an arming check for neutral throttle
-        ARMING_SKIP_CHECK_RPY = (1 << 6), // skip the an arming checks for the roll/pitch/yaw channels
-        ALLOW_SWITCH_REV      = (1 << 7), // honor the reversed flag on switches
+        IGNORE_RECEIVER         = (1 << 0), // RC receiver modules
+        IGNORE_OVERRIDES        = (1 << 1), // MAVLink overrides
+        IGNORE_FAILSAFE         = (1 << 2), // ignore RC failsafe bits
+        FPORT_PAD               = (1 << 3), // pad fport telem output
+        LOG_DATA                = (1 << 4), // log rc input bytes
+        ARMING_CHECK_THROTTLE   = (1 << 5), // run an arming check for neutral throttle
+        ARMING_SKIP_CHECK_RPY   = (1 << 6), // skip the an arming checks for the roll/pitch/yaw channels
+        ALLOW_SWITCH_REV        = (1 << 7), // honor the reversed flag on switches
+        CRSF_CUSTOM_TELEMETRY   = (1 << 8), // use passthrough data for crsf telemetry
```
```suggestion
        IGNORE_RECEIVER         = (1U << 0), // RC receiver modules
        IGNORE_OVERRIDES        = (1U << 1), // MAVLink overrides
        IGNORE_FAILSAFE         = (1U << 2), // ignore RC failsafe bits
        FPORT_PAD               = (1U << 3), // pad fport telem output
        LOG_DATA                = (1U << 4), // log rc input bytes
        ARMING_CHECK_THROTTLE   = (1U << 5), // run an arming check for neutral throttle
        ARMING_SKIP_CHECK_RPY   = (1U << 6), // skip the an arming checks for the roll/pitch/yaw channels
        ALLOW_SWITCH_REV        = (1U << 7), // honor the reversed flag on switches
        CRSF_CUSTOM_TELEMETRY   = (1U << 8), // use passthrough data for crsf telemetry
```

yaapu replied:
fixed

yaapu commented on libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp:
```diff
@@ -97,39 +105,48 @@ bool AP_Frsky_Telem::_set_telem_data(uint8_t frame, uint16_t appid, uint32_t dat
+/*
+  fetch Sport data for an external transport, such as FPort
+ */
+bool AP_Frsky_Telem::get_telem_data(uint8_t &frame, uint16_t &appid, uint32_t &data)
```
for sure, asking for a 1 size, I did ask myself the same, open to discussion I guess

yaapu replied:
I'll do this refactoring next will ping @peterbarker once done

yaapu commented on libraries/AP_Frsky_Telem/AP_Frsky_Telem.h:
```diff
@@ -58,15 +60,17 @@ class AP_Frsky_Telem {
 
+    // get next telemetry data for external consumers of SPort data (internal function)
+    bool _get_telem_data(uint8_t &frame, uint16_t &appid, uint32_t &data);
+    // return telemetry data for transports capable of bigger payloads (crossfire)
```
fixed

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -161,6 +171,21 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
+        uint8_t rf_mode;
+    } PACKED;
+
+    const volatile LinkStatus& get_link_status() const {
```
no cross-thread (in this use case) because the RCProtocol thread is calling the AP_RCTelemetry classes, I'll add a description

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -161,6 +171,21 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
         int8_t downlink_dnr; // ( db )
     } PACKED;
 
+    enum RFMode : uint8_t {
+        CRSF_RF_MODE_4HZ = 0,
+        CRSF_RF_MODE_50HZ,
+        CRSF_RF_MODE_150HZ,
+    };
```
This won't work because I have to compare it with link status data coming from a crossfire link status frame which uses an uint8_t for rf_mode, guess it could be made to work with a lot of casting

yaapu replied:
...done by using static_cast<>

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -161,6 +171,21 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
+
+    struct LinkStatus {
+        int16_t rssi = -1;
+        uint8_t rf_mode;
```
same as above

andyp1per replied:
I think this will work, but in the spec it's defined as uint8_t so I find that less error-prone

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+        return;
+    }
+
+    if (_custom_telem_init_done) {
```
yep, fixed

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LAT);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LON);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::TEXT);
+    passthrough->set_scheduler_entry_min_period(AP_Frsky_SPort_Passthrough::ATTITUDE, 120); // 5Hz
```
..easy answer, if I put 200 I don't get 100% 5Hz but rather 3.7-4.2 dependning on the board,, 120 gives between 5Hz and 6Hz

yaapu replied:
but since it's not critical let's make it 200 :-)

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    _custom_telem_init_done = true;
+}
+
+void AP_CRSF_Telem::update_custom_telemetry_rates(uint8_t rf_mode) {
```
see comment above abot rf_mode

yaapu replied:
done

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+    const uint8_t current_rf_mode = get_rf_mode();
```
see above

yaapu replied:
done

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    }
+    const uint8_t current_rf_mode = get_rf_mode();
+    if ( _telem_rf_mode != current_rf_mode) {
+        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d", _telem_rf_mode, current_rf_mode);
```
fixed

andyp1per commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -161,6 +171,21 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
+    struct LinkStatus {
+        int16_t rssi = -1;
+        uint8_t rf_mode;
+    } PACKED;
```
u r right, this is our structure so no need

yaapu commented:
tested on crossfire 4.06.
tested sport and fport get_telem_data refactoring.

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,140 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+        // to get crossfire firmware version we send an RX device ping until we get a response
+        // but only if there are no other requests pending
+        if (_crsf_version.pending && _pending_request.frame_type == 0) {
+            if (_crsf_version.retry_count++ > 50) {
```
Make this a constant somewhere

yaapu replied:
ok, done

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -245,7 +384,48 @@ void AP_CRSF_Telem::process_ping_frame(ParameterPingFrame* ping)
+        _crsf_version.use_rf_mode = true;
+    }
+
+    gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: fw version %d.%02d", _crsf_version.major, _crsf_version.minor);
```
Do we want this?

yaapu replied:
you mean do we really need to have crsf version info?
Well no, but it's cool to have it

yaapu replied:
thinking I can save a status text and send the version when I notify custom telemetry init done

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -837,12 +1113,11 @@ bool AP_CRSF_Telem::_get_telem_data(AP_RCProtocol_CRSF::Frame* data)
         return false;
     }
     memcpy(data->payload, &_telem, _telem_size);
-    data->device_address = AP_RCProtocol_CRSF::CRSF_ADDRESS_FLIGHT_CONTROLLER;  // sync byte
```
This isn't right. We are the source address.

yaapu replied:
...ok, fixed

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -173,13 +173,46 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
+        struct {
+            uint16_t appid;
+            uint32_t data;
+        } PACKED frames[9];
```
Maybe define a constant for this somewhere

yaapu replied:
done

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -228,33 +265,57 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
+        uint8_t retry_count;
+        bool use_rf_mode;
+        bool pending = true;
+    } PACKED _crsf_version;
```
Don't think this needs to be packed?

yaapu replied:
fixed

andyp1per commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -228,33 +265,57 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
+    struct {
+        uint8_t destination = AP_RCProtocol_CRSF::CRSF_ADDRESS_BROADCAST;
+        uint8_t frame_type;
+    } PACKED _pending_request;
```
Don't think this needs to be packed

yaapu replied:
fixed

andyp1per commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -1935,6 +1936,12 @@ void GCS::send_textv(MAV_SEVERITY severity, const char *fmt, va_list arg_list, u
+#if HAL_CRSF_TELEM_ENABLED
+    AP_CRSF_Telem* crsf = AP::crsf_telem();
+    if (crsf != nullptr) {
+        crsf->queue_message(severity, first_piece_of_text);
```
Does queue_message cope with when custom telemetry is disabled?

yaapu replied:
...very good point, we can skip pushing messages if not really needed

andyp1per reviewed (changes requested):
Looking good!

peterbarker commented:
On Wed, 18 Nov 2020, Alessandro Apostoli wrote:

> aren't we technically sending a ping and requesting device info?... i
> refactored the status text but not 100% convinced :-)

I didn't put together the challenge/response nature of hte protocol.  This 
renaming I was refering to was the fact we're requesting ourselves send 
something.

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -55,6 +55,9 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
         CRSF_FRAMETYPE_PARAMETER_WRITE = 0x2D,
         CRSF_FRAMETYPE_COMMAND = 0x32,
+        // Custom telemetry frame type is 0x80
+        CRSF_FRAMETYPE_CUSTOM_TELEM = 0x80,         // requires fw >= 4.06   
```
good point, it's reserved I'll rename it!

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -55,6 +55,9 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
         CRSF_FRAMETYPE_COMMAND = 0x32,
+        // Custom telemetry frame type is 0x80
+        CRSF_FRAMETYPE_CUSTOM_TELEM = 0x80,         // requires fw >= 4.06   
+        CRSF_FRAMETYPE_CUSTOM_TELEM_LEGACY = 0x7F,  // as suggested by Remo Masina for older fw
```
...and reorder it

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,140 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+    }
+    const AP_RCProtocol_CRSF::RFMode current_rf_mode = get_rf_mode();
+    if ( _telem_rf_mode != current_rf_mode) {
+        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d", (uint8_t)_telem_rf_mode, (uint8_t)current_rf_mode);
```
ok, added latest packet rate info like "CRSF: rf mode change %d->%d, rate is %dHz"

andyp1per replied:
Cool. So because of the flipping that people see I think you need to introduce some latency into this reporting to avoid annoying spam. For instance only print the message if the mode has changed and stay changed for longer than the equivalent RC timeout (e.g. 200ms)

yaapu replied:
yes, this would help, will do!

yaapu replied:
...mhmm not sure anymore, tested it and even with a 500ms grace period is still flipping between rf mode 1 and 2, I start to believe that is correct to let this pop to the user

yaapu replied:
this is the code I'm using
```
    const AP_RCProtocol_CRSF::RFMode current_rf_mode = get_rf_mode();
    const uint32_t now_ms = AP_HAL::millis();
    if ( _telem_rf_mode.mode != current_rf_mode && !_telem_rf_mode.change_notify_pending) {
        hal.console->printf("set: mode %d, time %lu\n", (uint8_t)_telem_rf_mode.mode, now_ms);
        _telem_rf_mode.last_change_time_ms = now_ms;
        _telem_rf_mode.change_notify_pending = true;
    } else if ((now_ms - _telem_rf_mode.last_change_time_ms) > 500 && _telem_rf_mode.change_notify_pending) {
        hal.console->printf("reset: mode %d, time %lu\n", (uint8_t)_telem_rf_mode.mode, now_ms);
        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d, rate is %dHz", (uint8_t)_telem_rf_mode.mode, (uint8_t)current_rf_mode, _scheduler.avg_packet_rate);
        update_custom_telemetry_rates(current_rf_mode);
        _telem_rf_mode.mode = current_rf_mode;
        _telem_rf_mode.change_notify_pending = false;
    }
```

andyp1per replied:
Yeah seems like quite a big delay, so maybe leave out until people complain! How come the rate is largely unchanged fomr mode 1 <-> mode 2?

yaapu replied:
...it doesn't change much because it's an average over 1sec, once it really switches to mode 1 it goes down (and that's the reason I did not print it before...it takes a while to settle on the real value)

yaapu replied:
I'll park this in a "stand by PR" :-)

yaapu commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -79,20 +80,140 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+
+void AP_CRSF_Telem::process_rf_mode_changes()
+{
+    if (!rc().crsf_custom_telemetry()) {
```
good point!

yaapu commented on libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h:
```diff
@@ -187,8 +215,8 @@ class AP_RCProtocol_CRSF : public AP_RCProtocol_Backend {
-    bool _fast_telem; // is 150Hz telemetry active
-    int16_t _current_rssi = -1;
+    
+    volatile struct LinkStatus _link_status;
```
the idea here was do not let the compiler optimize this struct

andyp1per replied:
Ok looks like this is equivalent to declaring each member as volatile which is ok I think

yaapu commented:
> On Wed, 18 Nov 2020, Alessandro Apostoli wrote: aren't we technically sending a ping and requesting device info?... i refactored the status text but not 100% convinced :-)
> I didn't put together the challenge/response nature of hte protocol. This renaming I was refering to was the fact we're requesting ourselves send something.

...I changed it to "requesting RX device info" which will make the both of us happy ;)

yaapu commented:
tested on 4.06

yaapu commented:
@andyp1per can you flight test this?

yaapu commented:
rebased and squashed

andyp1per commented:
@yaapu did a short test flight on my Y6B. The good news is it still flies and didn't crash, the bad news is almost nothing else appears to work for me. The widget thinks it has telemetry but all settings stay at 0. The crossfire configuration menu is unusably slow. I have updated the CRSF firmware to 4.08. So something seems crocked in the scheduler as it was working better before for me.

andyp1per commented:
With the new widget things working better. I am seeing

CRSF: rf mode change 1->2, rate is 56Hz
CRSF: rf mode change 1->2, rate is 57Hz
CRSF: rf mode change 1->2, rate is 58Hz
CRSF: rf mode change 1->2, rate is 59Hz

But never the other way round (2->1). I think probably you only care when the mode changes.
56Hz seems quite slow - I am on telem 3, is this the UART problem?

yaapu commented:
> With the new widget things working better. I am seeing
> 
> CRSF: rf mode change 1->2, rate is 56Hz
> CRSF: rf mode change 1->2, rate is 57Hz
> CRSF: rf mode change 1->2, rate is 58Hz
> CRSF: rf mode change 1->2, rate is 59Hz
> 
> But never the other way round (2->1). I think probably you only care when the mode changes.
> 56Hz seems quite slow - I am on telem 3, is this the UART problem?

Not all rf mode transitions are visible on OpenTX, only those from 1->2, this is because when in rf_mode 1 I discard all non debug/info status text messages to save bandwidth.
A 2->1 transition would trigger logging while in mode 1 (message discarded), a 1->2 transition triggers logging while in mode 2 (message sent).
To always see all transitions I could raise the severity of those messages to WARNING, just an idea.

On mode switches the reported telemetry rate is a moving average (last 4 samples) of the previous rate, so 56 from 1->2 could mean it was like 30Hz and growing. 
Now we know for sure (after testing with @andyp1per ) that non DMA UARTS do not support CRSF at full speed so yes, 56Hz was probably because UART3 did not have DMA.

yaapu commented:
tested on 4.06

tridge commented on libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp:
```diff
@@ -81,12 +81,12 @@ bool AP_Frsky_Telem::init(bool use_external_data)
-bool AP_Frsky_Telem::_get_telem_data(uint8_t &frame, uint16_t &appid, uint32_t &data)
+bool AP_Frsky_Telem::_get_telem_data(AP_Frsky_Backend::sport_packet_t* packet_array, uint8_t &packet_count, const uint8_t max_size)
 {
     if (_backend == nullptr) {
```
check packet_array == nullptr

yaapu replied:
fixed

tridge commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -245,7 +454,46 @@ void AP_CRSF_Telem::process_ping_frame(ParameterPingFrame* ping)
+    */
+    uint8_t offset = 0;
+    // get the terminator of the device name string
+    while (info->payload[offset] != 0x0 && offset++ < 42U);
```
use strnlen()

yaapu replied:
will do

tridge commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1090,86 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+    _telem_type = get_custom_telem_frame_id();
+    _telem.bcast.custom_telem.status_text.sub_type = AP_RCProtocol_CRSF::CustomTelemSubTypeID::CRSF_AP_CUSTOM_TELEM_STATUS_TEXT;
+    _telem.bcast.custom_telem.status_text.severity = _statustext.next.severity;
+    strncpy(_telem.bcast.custom_telem.status_text.text, _statustext.next.text, ARRAY_SIZE(_telem.bcast.custom_telem.status_text.text));
```
use strncpy_noterm()

yaapu replied:
will do

tridge commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -826,6 +1090,86 @@ void AP_CRSF_Telem::process_param_write_frame(ParameterSettingsWriteFrame* write
+    _telem.bcast.custom_telem.status_text.severity = _statustext.next.severity;
+    strncpy(_telem.bcast.custom_telem.status_text.text, _statustext.next.text, ARRAY_SIZE(_telem.bcast.custom_telem.status_text.text));
+    // add a potentially missing terminator
+    _telem.bcast.custom_telem.status_text.text[PASSTHROUGH_STATUS_TEXT_FRAME_MAX_SIZE-1] = 0;
```
use consistent length

yaapu replied:
will do

tridge commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -173,13 +178,46 @@ class AP_CRSF_Telem : public AP_RCTelemetry {
     } PACKED;
 
+    // Frame to hold passthrough telemetry
+    struct PassthroughSinglePacketFrame {
```
normal convention is "struct PACKED"

yaapu replied:
I'll fix other structs as well

atr2600 commented:
When is this pull request scheduled to be merged? 

I tested on a few different FCs and works amazing.

yaapu commented:
tested on crossfire 4.08:
- CRSF telemetry protocol @yaapu
- CRSF passthrough @yaapu
- LUA parameters download @yaapu
- CRSF on TRacer hardware @andyp1per
